File: backend/app/route_result_status.py
"""Top-level route result status contract for frontend consumption (V1C)."""
from __future__ import annotations

from enum import Enum

from backend.app.models import AgentState, HistoricalRoute


class RouteResultStatus(str, Enum):
    FULL_ROUTE = "FULL_ROUTE"
    PARTIAL = "PARTIAL"
    NO_ROUTE = "NO_ROUTE"
    ERROR = "ERROR"


_RUNTIME_ERROR_STATUSES = frozenset({"provider_error", "tool_failure"})


def _route_is_structurally_incomplete(route: HistoricalRoute, diagnostics: dict) -> bool:
    """True when the produced HistoricalRoute itself is incomplete."""
    if len(route.ordered_points) < 2:
        return True
    if route.branch_relations:
        return True
    if len(route.route_components) > 1:
        return True
    if route.route_components and not route.ordered_points:
        return True
    if int(diagnostics.get("contradictory_relation_count") or 0) > 0:
        return True
    return False


def _is_full_route(route: HistoricalRoute, diagnostics: dict) -> bool:
    return not _route_is_structurally_incomplete(route, diagnostics)


def derive_route_result_status(state: AgentState) -> RouteResultStatus | None:
    if state.requested_output != "historical_route":
        return None
    if state.status in _RUNTIME_ERROR_STATUSES:
        return RouteResultStatus.ERROR

    diagnostics = state.historical_route_diagnostics or {}
    route = state.historical_route

    if route is not None:
        if _is_full_route(route, diagnostics):
            return RouteResultStatus.FULL_ROUTE
        return RouteResultStatus.PARTIAL

    if state.historical_route_presentation:
        return RouteResultStatus.PARTIAL

    return RouteResultStatus.NO_ROUTE

File: frontend/src/route-result-status.ts
export type RouteResultStatus = "FULL_ROUTE" | "PARTIAL" | "NO_ROUTE" | "ERROR";

export function routeResultStatusLabel(status: RouteResultStatus | null | undefined): string | null {
  switch (status) {
    case "FULL_ROUTE":
      return "路线已生成";
    case "PARTIAL":
      return "部分路线 / 部分证据可用";
    case "NO_ROUTE":
      return "当前证据不足以生成可靠路线";
    case "ERROR":
      return "查询处理失败";
    default:
      return null;
  }
}

export function routeResultNoticeClass(status: RouteResultStatus | null | undefined): string | null {
  switch (status) {
    case "FULL_ROUTE":
      return "route-notice route-notice--full";
    case "PARTIAL":
      return "route-notice route-notice--partial";
    case "NO_ROUTE":
      return "route-notice route-notice--no-route";
    case "ERROR":
      return "route-notice route-notice--error";
    default:
      return null;
  }
}

export function shouldRenderRouteMap(
  status: RouteResultStatus | null | undefined,
  hasPresentation: boolean,
): boolean {
  return hasPresentation && (status === "FULL_ROUTE" || status === "PARTIAL");
}
